DRIVERS-3123 Relaxes requirement on reads of the ignored bits of PACKED_BIT vectors. (#1812)
mongodb/specifications: source/bson-binary-vector/bson-binary-vector.md

Drivers MUST validate vector metadata and raise an error if any invariant is violated:

- When unpacking binary data into a FLOAT32 Vector structure, the length of the binary data following the dtype and
    padding MUST be a multiple of 4 bytes.
- Padding MUST be 0 for all dtypes where padding doesn’t apply, and MUST be within [0, 7] for PACKED_BIT.
- A PACKED_BIT vector MUST NOT be empty if padding is in the range [1, 7].
    - For a PACKED_BIT vector with non-zero padding, ignored bits SHOULD be zero.

>         - When encoding, if ignored bits aren't zero, drivers SHOULD zero out or raise an exception, but drivers MAY leave

Reviewer: ```suggestion
        - When encoding, if ignored bits aren't zero, drivers SHOULD raise an exception, but drivers MAY leave
```

On further thought, I recommend raising an exception. Even in a major release, I am concerned zeroing out would be a quiet breaking change:

```python
# Server has vector stored with all ignored bits set to 1.
v = Binary.from_vector([0b11111111], BinaryVectorDtype.PACKED_BIT, padding=7) # Ignored bits are 1.
got = db.coll.find_one({"v": v}) # Might match depending on driver implementation.
```

Author: > On further thought, I recommend raising an exception. Even in a major release, I am concerned zeroing out would be a quiet breaking change:
> 
> ```python
> # Server has vector stored with all ignored bits set to 1.
> v = Binary.from_vector([0b11111111], BinaryVectorDtype.PACKED_BIT, padding=7) # Ignored bits are 1.
> got = db.coll.find_one({"v": v}) # Might match depending on driver implementation.
> ```

I prefer to error, too. :)

>         - When encoding, if ignored bits aren't zero, drivers SHOULD raise an exception, but drivers MAY leave them as-is if
>             backwards-compatibility is a concern.

Reviewer: ```suggestion
        - When encoding, if ignored bits aren't zero, drivers SHOULD raise an error, but drivers MAY leave them as-is for
            backwards-compatibility.
```
